backend/groth16/bls12-377/mpcsetup/setup_test.go

```diff
@@ -157,32 +176,317 @@ func (circuit *Circuit) Define(api frontend.API) error {
+
+	assertG1G2Equal(t, p1.parameters.G1.BetaTau[0], p1.parameters.G2.Beta)
+
+	commons := p1.Seal([]byte("beacon 1"))
+
+	for i := range commons.G2.Tau {
+		assertG1G2Equal(t, commons.G1.Tau[i], commons.G2.Tau[i])
+	}
+
+	assertG1G2Equal(t, commons.G1.BetaTau[0], commons.G2.Beta)
+}
+
+func TestCommonsUpdate(t *testing.T) {
+	t.Skip("skipping unit test for BLS12-377 so as not to clog the CI")
+	var c SrsCommons
+	c.setOne(1)
+	assertG1G2Equal(t, c.G1.BetaTau[0], c.G2.Beta)
+	one := fr.One()
+	var zero fr.Element
+	c.update(&zero, &zero, &one)
+	assertG1G2Equal(t, c.G1.BetaTau[0], c.G2.Beta)
+}
+
+func assertG1G2Equal(t *testing.T, p1 curve.G1Affine, p2 curve.G2Affine) {
+	_, _, g1, g2 := curve.Generators()
+	assertPairingsEqual(t, p1, g2, g1, p2)
+}
+
+// asserts e(p1, q1) = r(p2, q2)
+func assertPairingsEqual(t *testing.T, p1 curve.G1Affine, p2 curve.G2Affine, q1 curve.G1Affine, q2 curve.G2Affine) {
+	q1.Neg(&q1)
+	ok, err := curve.PairingCheck([]curve.G1Affine{p1, q1}, []curve.G2Affine{p2, q2})
+	require.NoError(t, err)
+	require.True(t, ok)
+}
+
+func TestPedersen(t *testing.T) {
+	t.Skip("skipping unit test for BLS12-377 so as not to clog the CI")
+	cs := getTestCircuit()
+	domainSize := ecc.NextPowerOfTwo(uint64(cs.GetNbConstraints()))
+
+	commons := commonsSmallValues(domainSize, 2, 3, 4)
+	var p Phase2
+	evals := p.Initialize(cs, &commons)
+	contributions := make([]fr.Element, 1+len(p.Sigmas))
+	for i := range contributions {
+		contributions[i].SetOne()
+	}
+	contributions[1].SetUint64(2)
+	p.update(&contributions[0], contributions[1:])
+	_, _, _, g2 := curve.Generators()
+	for i := range p.Sigmas {
+		assertPairingsEqual(t, evals.G1.CKK[0][i], p.Parameters.G2.Sigma[i], p.Parameters.G1.SigmaCKK[0][i], g2)
+	}
+}
+
+func TestPhase2Serialization(t *testing.T) {
+	t.Skip("skipping unit test for BLS12-377 so as not to clog the CI")
+	testRoundtrip := func(_cs constraint.ConstraintSystem) {
+		var (
+			p1 Phase1
+			p2 Phase2
+		)
+		p1.Initialize(ecc.NextPowerOfTwo(uint64(_cs.GetNbConstraints())))
+		commons := p1.Seal([]byte("beacon 1"))
+
+		p2.Initialize(_cs.(*cs.R1CS), &commons)
+		p2.Contribute()
+		require.NoError(t, gnarkio.RoundTripCheck(&p2, func() interface{} { return new(Phase2) }))
+	}
+
+	_cs, err := frontend.Compile(curve.ID.ScalarField(), r1cs.NewBuilder, &tinyCircuit{})
+	require.NoError(t, err)
+	testRoundtrip(_cs)
+
+	testRoundtrip(getTestCircuit())
+}
+
+type tinyCircuit struct {
+	X [4]frontend.Variable `gnark:",public"`
+}
+
+func (c *tinyCircuit) Define(api frontend.API) error {
+	for i := range c.X {
+		api.AssertIsEqual(c.X[i], i)
+	}
 	return nil
 }
 
-func (phase1 *Phase1) clone() Phase1 {
-	r := Phase1{}
-	r.Parameters.G1.Tau = append(r.Parameters.G1.Tau, phase1.Parameters.G1.Tau...)
-	r.Parameters.G1.AlphaTau = append(r.Parameters.G1.AlphaTau, phase1.Parameters.G1.AlphaTau...)
-	r.Parameters.G1.BetaTau = append(r.Parameters.G1.BetaTau, phase1.Parameters.G1.BetaTau...)
+func (p *Phase2) Equal(o *Phase2) bool {
```

Comment:
Is this method needed anymore when the roundtripchecker doesn't call `Equal` method?